std/range/package.d

```diff
@@ -948,134 +985,225 @@ if (Ranges.length > 0 &&
             }
             else
             {
-                @property bool empty()
+                @property bool empty() {return frontIndex >= backIndex;}
+            }
+
+            static if (allSatisfy!(isForwardRange, R))
+                @property auto save()
                 {
-                    foreach (i, Unused; R)
+                    auto saveI(size_t i)(){return source[i].save;}
+
+                    auto saveResult
+                        = Result(staticMap!(saveI, aliasSeqOf!(R.length.iota)));
+                    saveResult.frontIndex = frontIndex;
+                    static if (bidirectional) saveResult.backIndex = backIndex;
+                    return saveResult;
+                }
+
+            void popFront()
+            {
+                sw1: switch (frontIndex)
```

Comment:
Why do we need the switch alongside the costly `static foreach` when we can simply do `source[frontIndex].popFront`  and use `if` statements for the exceptional cases?

Reply:
We can't, since `source` is an alias sequence, not a random-access range. The index has to be known at compile time.

Reply:
:facepalm:

Reply:
Not a facepalm really, it's pretty much the reason why this range exists in the first place. The version where `source` is a range is called `joiner`.

Reply:
The facepalm was because I didn't figure it out and wasted your time.